src/shogun/machine/GLM.cpp

```diff
@@ -0,0 +1,254 @@
+	m_eta=2.0;
+}
+
+
+GLM::~GLM()
+{
+
+}
+
+std::shared_ptr<RegressionLabels> GLM::apply_regression(std::shared_ptr<Features> data)
+{
+	LinearMachine::set_features(std::static_pointer_cast<DotFeatures>(data));
+
+	if (!LinearMachine::features)
+		return std::make_shared<RegressionLabels>(SGVector<float64_t>());
+
+	int32_t num=LinearMachine::features->get_num_vectors();
+	ASSERT(num>0)
+	ASSERT(m_w.vlen==features->get_dim_feature_space())
+	SGVector<float64_t> out(num);
+	LinearMachine::features->dense_dot_range(out.vector, 0, num, NULL, m_w.vector, m_w.vlen, bias);
+	SGVector<float64_t> result = non_linearity(out);
+	return std::make_shared<RegressionLabels>(result);
+}
+
+void GLM::init_model(std::shared_ptr<Features> data)
+{
+	LinearMachine::set_features(std::static_pointer_cast<DotFeatures>(data));
+	SGMatrix<float64_t> X;
+	SGVector<float64_t> y;
+	NormalDistribution<float64_t> normal_dist;
+	auto n_features = LinearMachine::features->get_dim_feature_space();
+	
+	if (m_compute_bias && bias == 0)
+		bias = 1 / (n_features + 1) * normal_dist(m_prng);
+
+	if(!m_w.data())
+	{
+		m_w = SGVector<float64_t>(n_features);
+		for (int i = 0; i < n_features; i++)
+			m_w = 1 / (n_features + 1) * normal_dist(m_prng);
+	}
+}
+
+void GLM::iteration()
+{
+	SGMatrix<float64_t> X = LinearMachine::features->get_computed_dot_feature_matrix();
+	SGVector<float64_t> y = m_labels->get_values();
+	SGVector<float64_t> w_old(m_w);
+	SGVector<float64_t> gradient_w = compute_grad_L2_loss_w(X, y, m_w, bias);
+	float64_t gradient_bias = compute_grad_L2_loss_bias(X, y, m_w, bias);
+
+	//Update
+	m_w = linalg::add(m_w, gradient_w, 1.0, -1*m_learning_rate);
+	if(m_compute_bias)
+		bias = bias - m_learning_rate * gradient_bias;
+	
+	//Apply proximal operator
+	m_w = apply_proximal_operator(m_w, m_lambda * m_alpha);
+
+	//Convergence by relative parameter change tolerance
+	float64_t norm_update = linalg::norm(linalg::add(m_w, w_old, 1.0, -1.0));
+	if(m_current_iteration > 0 && (norm_update / linalg::norm(m_w)) < m_tolerance)
+		m_complete = true;
+}
+
+void GLM::set_tau(SGMatrix<float64_t> tau)
+{
+	m_tau = tau;
+}
+
+SGMatrix<float64_t> GLM::get_tau()
+{
+	return m_tau;
+}
+
+const SGVector<float64_t> GLM::conditional_intensity(const SGMatrix<float64_t> X, const SGVector<float64_t> w,const float64_t l_bias)
+{
+	SGVector<float64_t> z = compute_z(X, w, l_bias);
+	SGVector<float64_t> result = non_linearity(z);
+	return result;
+}
+
+const SGVector<float64_t> GLM::compute_z(const SGMatrix<float64_t> X, const SGVector<float64_t> w, const float64_t l_bias)
+{
+	SGVector<float64_t> z = linalg::matrix_prod(X, w, false);
+	return z;
+}
+
+const SGVector<float64_t> GLM::non_linearity(const SGVector<float64_t> z)
+{
+	SGVector<float64_t> result;
+	switch (distribution)
+	{
+	case POISSON:
+		result = SGVector<float64_t>(z);
+		float64_t l_bias = 0;
+
+		if(LinearMachine::m_compute_bias)
+			l_bias = (1 - m_eta) * std::exp(m_eta);
+
+		for (int i = 0; i < z.vlen; i++)
```

Comment:
Yes, I was looking for a way to do this using `linalg` but was unable to find a way. Here the value of the `result` vector depends on whether the corresponding value of `z` is more or less than `m_eta`. Is this possible using `linalg`?

This is the python code off which is this based:
```
result[z > m_eta] = z[z > m_eta] * np.exp(m_eta) + l_bias
result[z <= m_eta] = np.exp(z[z <= m_eta])
```

Reply:
dont think so.
@gf712 it is another thing we should have in a linear algebra structure: views for writing / reading. At least for reading, in jax, vetors are not mutable like this (not sure why, but one has to call a special assign function)

Reply:
Can I resolve this?

Reply:
you can but pls use a range based loop and no `int`

Reply:
Yes, I've done that, this is on an older commit so it isn't showing I guess.